### PR TITLE
Add functionality and tests for weighting samples in metrics

### PR DESCRIPTION
**Pull request by pechersky**
As described in #4313, it would be very useful in models where `sample_weight` is utilized to monitor accuracy and other metrics in a way that respects the weighing of the samples, either removing 0-weighted samples, removing some dimensions in temporal samples, or just scaling the vectors themselves.

This pull request adds this functionality by allowing metrics to be passed in as dicts with a`"metric"` key that holds the string (or function) value that would usually be the passed in metric, and a `"weighted"` key that should only hold the `"True"` value. This metric is wrapped in `weighted_metric`, allowing it to now take `sample_weight` values as well, similar to how `weighted_objective` works.

This pull request also includes tests for this new functionality, inspired by the weighted loss tests. Tests include verifying:
- new functionality does not disturb traditional metric calculation
- new functionality does not change metric values on trivial weights
- new functionality does change metric values when considering removing samples through 0 weighting
- new functionality works in traditional weighting_mode, as well as `temporal`, in both regular and "twist" format

### CODE REVIEW

File: tests/test_metric_weighting.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+def create_temporal_sequential_model():
+    model = Sequential()
+    model.add(GRU(32, input_shape=(timesteps, input_dim), return_sequences=True))
+    model.add(TimeDistributedDense(nb_classes))
+    model.add(Activation('softmax'))
+    return model
+
+
+# @keras_test
```
Comment (kilotaras) [inline tests/test_metric_weighting.py]: This line should probably be removed

File: tests/test_metric_weighting.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+                  nb_epoch=nb_epoch // 2, verbose=0,
+                  class_weight=class_weight, sample_weight=sample_weight)
+        model.fit(X_train, Y_train, batch_size=batch_size,
+                  nb_epoch=nb_epoch // 2, verbose=0,
+                  class_weight=class_weight, sample_weight=sample_weight,
+                  validation_split=0.1)
+
+
+# no evaluation weights: reference point
+model = create_sequential_model()
```
Comment (kilotaras) [inline tests/test_metric_weighting.py]: I think you want to all this in `setup_module`

http://doc.pytest.org/en/latest/xunit_setup.html

Comment (pechersky) [inline tests/test_metric_weighting.py]: I'll add a setup_module -- just for reference, `test_loss_weighting` produces the model in the top level of the module.

**Comment by pechersky** [comment]
I realized that `nonzero` is not doing what I expected to. Will require more work.
